bitcoin/src/pow.rs

```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+/// Error from `get_block_header_by_height` callback.
+///
+/// Occurse when there is an error while get a new block header
+/// from an external source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
+
+impl fmt::Display for GetBlockHeaderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            Self::Generic(error) => error,
+            Self::NotFound => "Block header not found",
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for GetBlockHeaderError {}
+
+/// Calculate the next Work required as the `GetNextWorkRequired` do
+/// in bitcoin core.
+///
+/// Return a error if the callback provide from the caller fails.
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensus_params: &Params,
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    mut get_block_header_by_height: F,
```

Comment:
I think many projects will desire `async` support here, can we get rid of the closure or something? IIUC this is just pulling all the blocks from the retarget period so we could just accept `IntoIterator<Item=I>, I: Borrow<Header>`. People would then just fetch all the blocks first and compute the difficulty afterwards. (I know it's a bit wasteful but I don't have a better idea that doesn't involve confusing API workarounds.) This also makes tests much easier.

Also if we do end up with closure we should use `Height` in the parameter.

Reply:
We don't need all the blocks; we only need the starting and ending blocks.

Reply:
So I guess there's an even easier soluton where we just take those headers directly as arguments :).

Reply:
If we only take the first and last block [header] we put the burden of doing this loop on the user

https://github.com/bitcoin/bitcoin/blob/82ea4e787c791acbc85fd3043dd6bae038cba4f2/src/pow.cpp#L32

FTR I don't actually understand what the checks are for in this loop.

Reply:
(And they have to understand that this loop is only required for testnet, so no one is going to do it.)

Reply:
>I think many projects will desire async support here, can we get rid of the closure or something?

Does the async support require also an async caller? otherwise, it is just a blocking call, right?

But I also like the idea of passing it as an input parameter, and the @tcharding is a good point.

This is the era of choice, we must support async or we do not want it at the API level right now? I leave you the fun of choose 😄 because I am bias